src/rvsdg/egglog_optimizer/rvsdg-logic.egg

```diff
@@ -0,0 +1,131 @@
+(datatype Situation
+    (always)
+    ; For last-iter, the Args of the predicate are bound in the Theta's context!
+    (last-iter Operand VecOperand VecOperand)
+
+    ; Some more situations we may want in the future:
+    ; (in-context Context)
+    ; ; For kth-iter, the Args of the predicate are bound in the surrounding
+    ; ; context!
+    ; ;         kth iter of this Theta  Theta's *outer* context
+    ; (kth-iter Body                    Context)
+
+    ; Note that all situations define the most recent binding body. We may want
+    ; situation to actually be a stack! E.g., we may want to reason about
+    ; an arg's value the kth iteration of a Theta, within a particular Gamma
+    ; branch
+    )
+
+(datatype Predicate
+    ; [[o1]]_{s1} <= [[o2]]_{s2}
+    ;    s1        o1      s2        o2
+    (leq Situation Operand Situation Operand)
+    ; We prefer to write eq with two leq's because it makes rules simpler
+    ; Some other predicates we may want:
+    ; (eq-mod-k i64 Situation Operand Situation Operand)
+    )
+
+
+(relation might-create-situation (Body Situation))
+(rule ((= body (Gamma pred inputs outputs)))
+      ((might-create-situation body (always)))
+      :ruleset fast-analyses)
+(rule ((= body (Theta pred inputs outputs)))
+      ((might-create-situation body (always)))
+      :ruleset fast-analyses)
+(rule ((= situation (last-iter pred inputs outputs))
+       (= body (Theta pred inputs outputs)))
+      ((might-create-situation body situation))
+      :ruleset fast-analyses)
+
+(relation holds (Predicate))
```

Comment:
We don't need holds- just use whether it's in the database or not
didn't catch this the first review sorry

Reply:
I considered this, but then simply constructing a Predicate (e.g. via `let`) would "make it true" - this feels wrong, and prevents us from having other relations over Predicates.